### PR TITLE
TF-3025 [SEARCH] Search result list auto refresh when perform some actions `read/star/...`

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#3025 

## Resolved

- Search result list auto refresh:

https://github.com/user-attachments/assets/45e91ff6-beb7-4891-9b9d-4b0b69a74677

- Actions for selection list incorrectly:

https://github.com/user-attachments/assets/e06074a6-0877-4389-893a-f903af70f332

### CODE REVIEW

**Review by tddang-linagora — APPROVED** [review]

**Comment by chibenwa** [comment]
If I do not search for unread emails then we do not need to update the search when the user reads a mail.

With the current implementation, would we refresh it?

**Comment by dab246** [comment]
> If I do not search for unread emails then we do not need to update the search when the user reads a mail.
> 
> With the current implementation, would we refresh it?

Yes, we will still refresh it. Please see below demo:


https://github.com/user-attachments/assets/1c0a4e2f-1947-4c97-93ae-e0cb3ec7f2d6

**Comment by chibenwa** [comment]
> Yes, we will still refresh it.

Ok so by doing this we just x10 the count of search requests to the server, correct?

**Comment by dab246** [comment]
> > Yes, we will still refresh it.
> 
> Ok so by doing this we just x10 the count of search requests to the server, correct?

No, I think I misunderstood you. 
1. If email is unread 
```
WHEN user open this email or click `blue dot` icon 
THEN refresh email list & email mark as read
```
2. If email is read 

```
WHEN user open this email
THEN not refresh email list
```

**Comment by hoangdat** [comment]
- [ ] when app has realtime update, search result also reload. We should not reload it

https://github.com/user-attachments/assets/b514c761-6b0a-4736-bf8d-5d62fb1d4c9c

**Comment by dab246** [comment]
> * [x]  when app has realtime update, search result also reload. We should not reload it
> 
>  Screen.Recording.2024-10-09.at.09.13.38.mov

It still persists the list when real time update. Please check again.



https://github.com/user-attachments/assets/086a81da-6732-4f82-8858-781f09a49ac7

File: lib/features/thread/presentation/thread_controller.dart [hunk 1]
```diff
@@ -810,12 +814,14 @@ class ThreadController extends BaseController with EmailActionController {
     searchController.clearTextSearch();
   }
 
-  void _searchEmail({UnsignedInt? limit}) {
+  void _searchEmail({UnsignedInt? limit, bool isRefreshChange = false}) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:817]:
```suggestion
  void _searchEmail({UnsignedInt? limit, bool needRefreshSearchState = false}) {
```
WDYT?

File: lib/features/thread/presentation/thread_controller.dart [hunk 2]
```diff
@@ -585,7 +589,7 @@ class ThreadController extends BaseController with EmailActionController {
   void _refreshEmailChanges({jmap.State? currentEmailState}) {
     log('ThreadController::_refreshEmailChanges(): currentEmailState: $currentEmailState');
     if (searchController.isSearchEmailRunning) {
-      _searchEmail(limit: limitEmailFetched);
+      _searchEmail(limit: limitEmailFetched, isRefreshChange: true);
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:592]: we can write the test:
```
WHEN thread controller in searching, 
AND `MarkAsStarEmailSuccess` is comming,
THEN `_searchEmailInteractor` is invoked with proper arguments
AND `listEmailController` should not jumped
AND `mailboxDashBoardController.emailsInCurrentMailbox` should not be cleared
```

File: lib/features/thread/presentation/thread_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/thread/presentation/thread_controller.dart:817]: it doesn't make sense in this case

**Comment by hoangdat** [comment]
Please test mobile UI too 
- [ ] selection appbar is not display well 
IMG_6435

File: lib/features/thread/domain/usecases/search_email_interactor.dart [hunk 3]
```diff
@@ -26,10 +26,15 @@ class SearchEmailInteractor {
       Set<Comparator>? sort,
       Filter? filter,
       Properties? properties,
+      bool isRefreshChange = false,
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/thread/domain/usecases/search_email_interactor.dart:29]: done

File: lib/features/thread/presentation/thread_controller.dart (hunk 2 above)
Comment (dab246) [reply lib/features/thread/presentation/thread_controller.dart]: done

**Review by tddang-linagora — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**